velox/connectors/fuzzer/tests/FuzzerConnectorTest.cpp

```diff
@@ -0,0 +1,163 @@
+    }
+    return splits;
+  }
+
+  std::shared_ptr<FuzzerTableHandle> makeFuzzerTableHandle(
+      size_t fuzzerSeed = 0) const {
+    return std::make_shared<FuzzerTableHandle>(
+        kFuzzerConnectorId, fuzzerOptions_, fuzzerSeed);
+  }
+
+ private:
+  VectorFuzzer::Options fuzzerOptions_;
+};
+
+TEST_F(FuzzerConnectorTest, singleSplit) {
+  const size_t numRows = 100;
+  auto type = ROW({BIGINT(), DOUBLE(), VARCHAR()});
+
+  auto plan =
+      PlanBuilder().tableScan(type, makeFuzzerTableHandle(), {}).planNode();
+
+  exec::test::AssertQueryBuilder(plan)
+      .split(makeFuzzerSplit(numRows))
+      .assertNumRows(type, numRows);
+}
+
+TEST_F(FuzzerConnectorTest, complexTypes) {
+  const size_t numRows = 100;
+  auto type = ROW({
+      ARRAY(BIGINT()),
+      ROW({VARCHAR(), MAP(INTEGER(), ARRAY(DOUBLE())), VARBINARY()}),
+      REAL(),
+  });
+
+  auto plan =
+      PlanBuilder().tableScan(type, makeFuzzerTableHandle(), {}).planNode();
+
+  exec::test::AssertQueryBuilder(plan)
+      .split(makeFuzzerSplit(numRows))
+      .assertNumRows(type, numRows);
+}
+
+TEST_F(FuzzerConnectorTest, multipleSplits) {
+  const size_t rowsPerSplit = 100;
+  const size_t numSplits = 10;
+  auto type = ROW({BIGINT(), DOUBLE(), VARCHAR()});
+
+  auto plan =
+      PlanBuilder().tableScan(type, makeFuzzerTableHandle(), {}).planNode();
+
+  exec::test::AssertQueryBuilder(plan)
+      .splits(makeFuzzerSplits(rowsPerSplit, numSplits))
+      .assertNumRows(type, rowsPerSplit * numSplits);
+}
+
+TEST_F(FuzzerConnectorTest, randomTypes) {
+  const size_t rowsPerSplit = 100;
+  const size_t numSplits = 10;
+
+  const size_t iterations = 20;
+
+  for (size_t i = 0; i < iterations; ++i) {
+    auto type = VectorFuzzer({}, pool()).randRowType();
+
+    auto plan =
+        PlanBuilder().tableScan(type, makeFuzzerTableHandle(), {}).planNode();
+    exec::test::AssertQueryBuilder(plan)
+        .splits(makeFuzzerSplits(rowsPerSplit, numSplits))
+        .assertNumRows(type, rowsPerSplit * numSplits);
+  }
+}
+
+TEST_F(FuzzerConnectorTest, reproducible) {
+  const size_t numRows = 100;
+  auto type = ROW({BIGINT(), ARRAY(INTEGER()), VARCHAR()});
+
+  auto plan1 = PlanBuilder()
+                   .tableScan(type, makeFuzzerTableHandle(/*seed=*/1234), {})
+                   .planNode();
+  auto plan2 = PlanBuilder()
+                   .tableScan(type, makeFuzzerTableHandle(/*seed=*/1234), {})
+                   .planNode();
+
+  auto results1 = exec::test::AssertQueryBuilder(plan1)
+                      .split(makeFuzzerSplit(numRows))
+                      .copyResults(pool());
+  auto results2 = exec::test::AssertQueryBuilder(plan2)
+                      .split(makeFuzzerSplit(numRows))
+                      .copyResults(pool());
+
+  exec::test::assertEqualResults({results1}, {results2});
+}
+
+} // namespace
+
+int main(int argc, char** argv) {
+  testing::InitGoogleTest(&argc, argv);
+  folly::init(&argc, &argv, false);
+  return RUN_ALL_TESTS();
+}
```

Comment:
Curious why we create an explicit main function here? I found some of our tests have "main" and some do not. Is there a reason behind it?

Reply:
I copied from another unit test. I remember there was about that if you don't have a main explicitly defined, gtest will give you one, but that one won't call folly::init. And that broke folly singleton initialization in some cases.